### PR TITLE
add Euler's Constant

### PR DESCRIPTION
**Pull request by abhinavagarwal07**
added symbolic constant `euler_const`
#697

### CODE REVIEW

**Comment by certik** [comment]
We should add a test for it, and at least some evaluation, e.g. double
precision (using eval double visitor).

File: symengine/constants.cpp [hunk 1]
```diff
@@ -46,6 +46,7 @@ RCP<const Number> I = Complex::from_two_nums(*zero, *one);
 
 RCP<const Constant> pi = constant("pi");
 RCP<const Constant> E = constant("E");
+RCP<const Constant> euler_const = constant("euler_const");
```
Comment (isuruf) [inline symengine/constants.cpp]: Name of this constant should be changed. Maybe `EulerGamma` as in SymPy and Mathematica or `euler_gamma` as in Sage

File: symengine/eval_arb.cpp [hunk 2]
```diff
@@ -286,6 +286,8 @@ class EvalArbVisitor : public BaseVisitor<EvalArbVisitor> {
             arb_const_pi(result_, prec_);
         } else if (x.__eq__(*E)) {
             arb_const_e(result_, prec_);
+        } else if(x.__eq__(*euler_const)) {
+            arb_const_euler_const(result_, prec_);
```
Comment (isuruf) [inline symengine/eval_arb.cpp]: should be `arb_const_euler`

**Comment by isuruf** [comment]
Can you also add numerical evaluation for `eval_mpfr` and `eval_mpc` using `mpfr_const_euler`?

**Comment by abhinavagarwal07** [comment]
A doubt 
http://www.mpfr.org/mpfr-2.3.0/mpfr.html#Special-Functions
mpfr_const_euler corresponds to value of Euler’s constant 0.577
however `eval_mpc.cpp` uses `mpfr_const_euler` when x is E.

**Comment by isuruf** [comment]
`eval_mpc.cpp` should be corrected

**Comment by abhinavagarwal07** [comment]
Also `eval_mpfr.cpp` uses `mpfr_const_euler` when x is E

File: symengine/tests/eval/test_eval_double.cpp [hunk 3]
```diff
@@ -58,7 +59,7 @@ TEST_CASE("eval_double: eval_double", "[eval_double]")
     r2 = sin(div(integer(1), integer(2)));
     r3 = div(one, integer(5));
     r4 = integer(5);
-
+
```
Comment (isuruf) [inline symengine/tests/eval/test_eval_double.cpp]: Remove this extra space

**Comment by isuruf** [comment]
+1 to merge after above is fixed.

File: symengine/eval_double.cpp [hunk 4]
```diff
@@ -197,6 +197,8 @@ class EvalDoubleVisitor : public BaseVisitor<U> {
             result_ = std::atan2(0, -1);
         } else if (eq(x, *E)) {
             result_ = std::exp(1);
+        } else if (eq(x, *EulerGamma)) {
+            result_ = 0.5772156649015328606065; //use until polygamma or digamma is not implemented
```
Comment (certik) [inline symengine/eval_double.cpp]: `//use until polygamma or digamma is implemented` ? (Remove `not`, add space after `//`.)

File: symengine/eval_double.cpp [hunk 5]
```diff
@@ -420,6 +422,8 @@ std::vector<fn> init_eval_double()
             return ::atan2(0, -1);
         } else if (eq(x, *E)) {
             return ::exp(1);
+        } else if (eq(x, *EulerGamma)) {
+            return 0.5772156649015328606065; //use until polygamma or digamma is not implemented
```
Comment (certik) [inline symengine/eval_double.cpp]: Same here.

**Comment by certik** [comment]
Also add tests for `mpc` and `arb`. The tests for `double` and `mpfr` are fine.

Otherwise it looks good.

**Comment by certik** [comment]
Thanks for fixing the comments. Please ping me once you add tests for `mpc` and `arb`, or if you need any help, let me know.

**Comment by abhinavagarwal07** [comment]
@certik 
added tests for `arb`, but I need suggestions for `mpc`.

File: symengine/tests/eval/test_eval_arb.cpp [hunk 6]
```diff
@@ -672,3 +675,28 @@ TEST_CASE("Coth: eval_arb", "[eval_arb]")
...
+    mpfr_t f;
+    mpfr_init2(f, 57);
+    eval_mpfr(f, *r1, MPFR_RNDN);
+
+    REQUIRE(arb_contains_mpfr(a, f));
+    
+    eval_arb(a, *r1, 45);
+    eval_mpfr(f, *r1, MPFR_RNDN);
+
+    REQUIRE(arb_contains_mpfr(a, f));
```
Comment (isuruf) [inline symengine/tests/eval/test_eval_arb.cpp]: Above line checks the same as in line 693 right?

File: symengine/tests/eval/test_eval_arb.cpp [hunk 7]
```diff
@@ -672,3 +675,28 @@ TEST_CASE("Coth: eval_arb", "[eval_arb]")
...
     arb_clear(a);
 }
+
+TEST_CASE("Constants: eval_arb", "[eval_arb]")
+{
+    arb_t a;
+    arb_init(a);
+    
+    RCP<const Basic> r1 = mul(EulerGamma, integer(100000000));
+    RCP<const Basic> r2 = div(sub(r1, integer(57721566)),integer(100000000));
```
Comment (isuruf) [inline symengine/tests/eval/test_eval_arb.cpp]: Space after comma

File: symengine/tests/eval/test_eval_double.cpp [hunk 8]
```diff
@@ -71,7 +71,10 @@ TEST_CASE("eval_double: eval_double", "[eval_double]")
         { add(sinh(one), add(cosh(one), add(tanh(one), coth(one)))), 4.792911269914 },
         { add(asinh(r4), add(acosh(r4), add(atanh(r3), acoth(r4)))), 5.010335118942 },
         { SymEngine::abs(log(div(pi, mul(E, integer(2))))), 0.548417294710 },
-        { SymEngine::atan2(r1, neg(r2)), 2.08867384922582 }
+        { SymEngine::atan2(r1, neg(r2)), 2.08867384922582 },
+        { mul(pi,mul(E,EulerGamma)), 4.92926836742289 },
```
Comment (isuruf) [inline symengine/tests/eval/test_eval_double.cpp]: Space after comma

File: symengine/tests/eval/test_eval_double.cpp [hunk 9]
```diff
@@ -71,7 +71,10 @@ TEST_CASE("eval_double: eval_double", "[eval_double]")
...
+        { pow(mul(EulerGamma, r4), integer(8)), 4813.54354505117582 },
+        { mul(EulerGamma ,integer(10)), 5.7721566490153286 }
```
Comment (isuruf) [inline symengine/tests/eval/test_eval_double.cpp]: Should be `Eulergamma, integer(10)`

**Comment by isuruf** [comment]
For `mpc` add a test for something like `1 + Eulergamma * I`

File: symengine/tests/eval/test_eval_arb.cpp (hunk 6 above)
Comment (abhinavagarwal07) [inline symengine/tests/eval/test_eval_arb.cpp]: I think so.

**Comment by abhinavagarwal07** [comment]
@isuruf
Thanks for suggestions and corrections

File: symengine/tests/eval/test_eval_arb.cpp (hunk 6 above)
Comment (isuruf) [inline symengine/tests/eval/test_eval_arb.cpp]: Can you remove these lines 695-698?

Comment (certik) [inline symengine/tests/eval/test_eval_arb.cpp]: Yes, looks like those should be removed.

**Comment by certik** [comment]
@abhinavagarwal07 can you please fix the last comment by @isuruf: https://github.com/symengine/symengine/pull/700#discussion_r48531845, otherwise this looks good.

**Comment by abhinavagarwal07** [comment]
```
CMake Error at C:/Program Files (x86)/CMake/share/cmake-3.4/Modules/FindPackageHandleStandardArgs.cmake:148 (message):
  Could NOT find GMP (missing: GMP_LIBRARIES GMP_INCLUDE_DIRS)
```
seems like it's appveyor's error.

**Comment by abhinavagarwal07** [comment]
@certik @isuruf 
done

**Comment by certik** [comment]
Thanks for fixing it! Looks good to me know.

**Pull request merged**